README.md
Yandex.Money
================================
Yandex.Money is a joint venture of Sberbank, the largest bank in Eastern Europe, and the Internet company Yandex, operator of Russia's leading search engine. Yandex.Money is the easiest way to start sales in Russia. We'll provide your online store with the means of accepting all payment options popular in Russia — cash, bank cards, e‑money, and mobile carrier billing.

## Yandex.Kassa  [**RUS**](https://money.yandex.ru/start/#1) [**ENG**](https://money.yandex.ru/eng/merchants/)

**Supported frameworks**
> * [**Django**](https://github.com/aTastyCookie/yandexmoney_django) - Django is a high-level Python Web framework that encourages rapid development and clean, pragmatic design.
> * [**Omnipay**](https://github.com/aTastyCookie/yandexmoney_omnipay) is a payment processing library for PHP. It has been designed based on ideas from Active Merchant, plus experience implementing dozens of gateways for CI Merchant. It has a clear and consistent API, is fully unit tested, and even comes with an example application to get you started.

**Supported CMS - YandexModule** (YandexKassa, YandexMoney, YandexMetrika,YandexMarket.CPC, YandexMarket.CPA)
> * [**PrestaShop**](https://github.com/aTastyCookie/yandex_prestashop) free ecommerce solution to start your online business and start selling online
> * [**OpenCart 2**](https://github.com/aTastyCookie/yandex_opencart2) - free shopping cart system. OpenCart is an open source PHP-based online e-commerce solution.

**Supported CMS - YandexMoney**
> * [**1С-Битрикс**](http://dev.1c-bitrix.ru/learning/course/index.php?COURSE_ID=42&LESSON_ID=3547) is the industry's most secure website management, e-commerce and e-marketing solution that delivers extreme value in performance, functionality, and ease-of-use for small and medium-sized businesses.
> * [**Merchium**](http://www.merchium.ru/yandex/) - powerful and free eCommerce site builder, officially approved by Yandex.Money. Merchium lets you create a ready-to-sell store in 10 seconds and embed it any existing site, blog, or Facebook page.
> * [**nethouse**](http://nethouse.ru/) - free website builder, build a website free
> * **UMI** professional Content Management System
> * * [RU](http://umi.ru/) - create site for free
> * * [CMS](http://www.umi-cms.ru/) - CMS new generation
> * [**InSales**](http://www.insales.ru/) - Online Platform - Support 24x7 - Running without programmer - Finished design and integration
> * [**OpenCart**](https://github.com/aTastyCookie/yandexmoney_opencart) - free shopping cart system. OpenCart is an open source PHP-based online e-commerce solution.
> * [**CS-Cart**](http://yandexmoney.cs-cart.ru/instructions.html) - feature-rich eCommerce software that powers over 30 000 stores around the globe.
> * [**X-Cart**](http://www.x-cart.ru/yandex-money.html) - online Shopping cart software & PHP ecommerce solutions for the best ecommerce websites: X-Cart.
> * [**osCommerce**](https://github.com/aTastyCookie/yandexmoney_oscommerce) - (styled "osCommerce" - "open source Commerce") is an e-commerce and online store-management software program. 
> * **ModX** is the web content management platform for those that truly care about no-compromise design and exceptional user experience. It gives you complete control over your site and content, with the flexibility and scalability to adapt to your changing needs.
> * * [Revo](https://github.com/aTastyCookie/yandexmoney_modx)
> * * [Evo](https://github.com/aTastyCookie/yandexmoney_modx_evo)
> * [**WebAsyst Shop-Script-5**](http://www.shop-script.ru/) - feature-rich PHP shopping cart solution & ecommerce framework developed in 2013-2014. One of the leading shopping cart solutions of Russia. Over 2500 live stores.
> * **WordPress** is web software you can use to create a beautiful website or blog. We like to say that WordPress is both free and priceless at the same time.
> * * [WP-Shop](https://wordpress.org/plugins/wp-shop-yandex/)
> * * [eShop](https://github.com/aTastyCookie/yandexmoney_wp_eshop)
> * * [WooCommerce](https://github.com/aTastyCookie/yandexmoney_wp_woocommerce)
> * [**AdvantsShop.NET**](http://www.advantshop.net/) - easy and intuitive engine online store for a quick start of your business. Open your own shop now!
> * [**AdvantsShop.NET**](http://www.advantshop.net/) - easy and intuitive engine online store for a quick start of your business. Open your own shop now!
> * [**Megagroup.ru CMS.S3**](http://megagroup.ru/cms) - сreation of website.
> * [**Fast-Sales**](http://fast-sales.ru/) - ready CMS to run an online store for 1 day
> * [**Magento**](http://www.magentocommerce.com/magento-connect/catalog/product/view/id/19716/s/yandex-money-payment-module/category/12848/)  is the eCommerce software and platform trusted by the world's leading brands. Grow your online business with Magento!
> * [**Rugento**](http://www.rugento.ru/yandexmoney-payment-module.html) - platform for the growth of your business!
> * [**Difocus**](http://shopexpress.difocus.ru/new_help/payments/yakassa) a turnkey solution to build your online store
> * [**UCOZ**](http://manual.ucoz.net/board/34-1-0-549) free content management system and hosting for websites created with its use.
> * [**Payplug.In**](http://payplug.in/) technical integrator payments
> * **Joomla** is the mobile-ready and user-friendly way to build your website.
> * * [VirtueMart](https://github.com/aTastyCookie/yandexmoney_joomla_mammuthus)
> * * [JoomShopping](https://github.com/aTastyCookie/yandexmoney_joomla) & [JoomShopping PRO](http://joomshopping.pro/yandex-cashbox.html)
> * [**VamShop**](http://vamshop.ru/) a turnkey solution to build your online store
> * [**NetCat**](http://netcat.ru/) content management system of any complexity
> * [**HostCMS**](https://github.com/aTastyCookie/yandexmoney_hostcms) сomfortable and functional engine for website and e-shop, CMS with great potential
> * [**ABCP**](http://docs.abcp.ru/wiki/Payments:YandexMoney) a platform for online auto parts store
> * [**SimpaCMS**](http://simplacms.ru/) a turnkey solution to build your online store
> * [**SiteEdit**](http://help.siteedit.ru/yandex-kassa/) this is the most clear and easy to use program for creating sites that rightfully gained the attention of thousands of users.
> * [**Diafan**](http://cms.diafan.ru/) a turnkey solution to build your online store
> * [**ShopCMS**](https://github.com/aTastyCookie/yandexmoney_shopcms) is a professional online store script that allows to organize e-commerce at the highest level with minimal cost.
> * [**Автоофис**](https://autoweboffice.com/) a platform for doing business online.
> * [**GetShopApp**](https://www.getshopapp.com/) service creating a mobile version of online stores for iPhone and Android
> * [**ShopWare**](https://github.com/aTastyCookie/yandexmoney_shopware) is a powerful and flexible eCommerce software. Innovation, advanced technologies and comprehensive services make Shopware one of the most promising eCommerce solutions on the market.
> * [**MegaGroup**](http://megagroup.ru/) professional content management system
> * [**PrestaShop**](https://github.com/aTastyCookie/yandex_prestashop) free ecommerce solution to start your online business and start selling online
> * [**Drupal**](https://github.com/aTastyCookie/yandexmoney_drupal_commerce_and_ubercart) is a free and open-source content-management framework written in PHP and distributed under the GNU General Public License.
> * [**ECShop**](https://github.com/aTastyCookie/yandexmoney_ecshop) very popular in China script online store.
> * [**Лейка**](https://github.com/Teplitsa/Leyka) is a free plug-in (independent software module) for WordPress (the system that is running a lot of web sites), allowing to organize fundraising for your organization or individual initiative.
> * [**n4.biz**](http://n4.biz/recieve-payments) business social network B2B and B2C. Create an online store for free
> * [**Resonline**](http://resonline.ru/) online platform to automate processes and increase profits of your hotel
> * [**HubPay**](http://hubpay.ru/) & [**HubPay FB**](https://www.facebook.com/hubpayinc) CRM-system for professional billing and customer service 
> * [**PayByLink**](http://paybylink.com/instruction/) & [**PayByLink FB**](https://www.facebook.com/paybylink) invoice manager

**Example**
> * [**PHP**](https://github.com/aTastyCookie/YandexMoney/tree/master/Yandex.Kassa/example%20integration/php) - Example in PHP

## Yandex.Money.API

**Documentation**
> * [API Денег для приложений](https://tech.yandex.ru/money/) - russian
> * [Yandex.Money API](https://tech.yandex.com/money/) - english

## Yandex.Money.Payout

**Documentation**
> * [Выплатной протокол](https://github.com/aTastyCookie/YandexMoney/tree/master/Yandex.Money.Payout/documentation/protocol/RUS) - russian
> * [Payout protocol](https://github.com/aTastyCookie/YandexMoney/tree/master/Yandex.Money.Payout/documentation/protocol/ENG) - english


CHANGELOG.md
# PROTOCOL STATUS

###paymentTypes 
* paymentType=PC - yandexWallet
* paymentType=AC - creditCard
* paymentType=GP - kiosks
* paymentType=MC - mobileCommerce
* paymentType=WM - webMoney
* paymentType=SB - sberbankOnline - bank invoicing
* paymentType=AB - alphaClick - bank invoicing

###protocols
* [commonHTTP3.x](https://github.com/aTastyCookie/YandexMoney/tree/master/Yandex.Kassa/documentation/protocol)
* [MWS 3.x](https://github.com/aTastyCookie/YandexMoney/tree/master/Yandex.Kassa/documentation/protocol)
* [payoutProtocol](https://github.com/aTastyCookie/YandexMoney/tree/master/Yandex.Money.Payout/documentation/protocol)
* [API](https://tech.yandex.ru/money/)


php/refund.php
 <?php

    protected function encrypt($data, $certFile, $keyFile)
    {
        try {
            $pipes = array();
            $process = proc_open(
                'openssl smime -sign -signer ' . $certFile . ' -inkey ' . $keyFile . ' -nochain -nocerts -outform PEM -nodetach',
                array(array("pipe", "r"), array("pipe", "w"), array("pipe", "w")),
                $pipes
            );

            if (is_resource($process)) {
                fwrite($pipes[0], $data);
                fclose($pipes[0]);
                $pkcs7 = stream_get_contents($pipes[1]);
                fclose($pipes[1]);
                proc_close($process);
                return $pkcs7;
            }
            return false;
        } catch (Exception $e) {
            return false;
        }
    }

    public function sendRefundRequest($cause)
    {
        $data = '<?xml version="1.0" encoding="UTF-8"?>
<returnPaymentRequest clientOrderId="' . $this->id . '" requestDT="' . date('Y-m-d\TH:i:sP') . '"
invoiceId="' .$this->invoiceId . '" shopID="' .$this->shopId . '" amount="' .sprintf("%.2f", $this->cost) . '"
currency="10643" cause="' .$cause . '" />';

        $certFile = '/path_to_project/ssl/cert.cer';
        $keyFile = '/path_to_project/ssl/key.key';
        if (($data = $this->encrypt($data, $certFile, $keyFile)) === false)
            return false;

        $ch = curl_init();
        curl_setopt($ch, CURLOPT_URL, 'https://penelope-demo.yamoney.ru:8083/webservice/mws/api/returnPayment');
        curl_setopt($ch, CURLOPT_HTTPHEADER, array('Content-Type:application/x-www-form-urlencoded'));
        curl_setopt($ch, CURLOPT_POST, true);
        curl_setopt($ch, CURLOPT_POSTFIELDS, $data);
        curl_setopt($ch, CURLOPT_VERBOSE, false);
        curl_setopt($ch, CURLOPT_HEADER, 0);
        curl_setopt($ch, CURLOPT_RETURNTRANSFER, true);
        curl_setopt($ch, CURLOPT_CONNECTTIMEOUT, 30);
        curl_setopt($ch, CURLOPT_USERAGENT, 'Ymoney CollectMoney');
        curl_setopt($ch, CURLOPT_SSL_VERIFYPEER, false);
        curl_setopt($ch, CURLOPT_SSL_VERIFYHOST, false);
        curl_setopt($ch, CURLOPT_SSLCERT, $certFile);
        curl_setopt($ch, CURLOPT_SSLKEY, $keyFile);
        curl_setopt($ch, CURLOPT_SSLCERTPASSWD, $this->_pswd);
        curl_setopt($ch, CURLOPT_FORBID_REUSE, TRUE);
        curl_setopt($ch, CURLOPT_FRESH_CONNECT, TRUE);
        $response = curl_exec($ch);
        curl_close($ch);

        header('Content-Type:text/xml');
        die($response);
    }

?>


Yandex.Kassa/example integration/php/yapay.php
<?
namespace Yapay;

class Payment
{
    const SHOP_ID = 666;

    const USERAGENT = 'useragent';

    const YMAPI = 'https://penelope-demo.yamoney.ru:8083/webservice/mws/api/';

    const REPEAT_CARD_PAYMENT = 'repeatCardPayment';

    const LISTORDERS = 'listOrders';

    const HTTPHEADER = 'Content-Type:application/x-www-form-urlencoded';

    const CERTFILE = 'youcert.cer';

    const KEYFILE = 'private.key';

    const SERTPASS = 'sertpass';

    const TMPLOGFILE = './._temp_curl_log.txt';

    /**
     * @var Error переменная содержит название ошибки последней операции
     */
    public $error = NULL;

    /**
     * @var string имя последнего вызванного метода
     */
    public $lastMethod = "";

    /**
     * @var SimpleXMLElement содержит xml-объект последнего ответа
     */
    public $lastXML = NULL;

    /**
     * @var string
     */
    public $curlTrace = null;

    /**
     * @var bool флаг включает/выключает логирование выполнения curl запросов и ошибок в файл
     */
    private $debug = TRUE;

    /**
     * @var array содержит последние параметры запроса
     */
    private $lastRequestData = "";

    /**
     * @var string содержит кэш последнего ответа
     */
    public  $lastResponse = "";

    /**
     * @var указатель на файл для записи лога выполнения запроса через curl
     */
    private $logout = null;

    public function getLogfilePath()
    {
        return dirname(__FILE__) . '/../logs/yapay-log.txt';
    }

    public function getMethodUrl($name)
    {
        return self::YMAPI . $name;
    }

    public function repeatCardPayment($clientOrderID, $amount, $orderNumber, $invoiceId)
    {
        $requestData = array(
            'clientOrderId' => $clientOrderID,
            'invoiceId' => $invoiceId,
            'amount' => $amount,
            'orderNumber' => $orderNumber
        );
        return $this->doRequest(self::REPEAT_CARD_PAYMENT, $requestData);
    }

    public function listOrders($options)
    {
        $now = new DateTime('NOW');
        $requestData = array(
            "shopId" => self::SHOP_ID,
            "requestDT" => $now->format(DateTime::ISO8601),
        );

        if ($options["orderNumber"]) {
            $requestData["orderNumber"] = $options["orderNumber"];
        }

        if ($options["dateFrom"]) {
            $requestData['paymentDatetimeGreaterOrEqual'] = $options['dateFrom'];
        }

        if ($options["dateTo"]) {
            $requestData['paymentDatetimeLessOrEqual'] = $options['dateTo'];
        }

        return $this->doRequest(self::LISTORDERS, $requestData);
    }

    private function initCurl($url, $data = null)
    {
        $ch = curl_init();
        curl_setopt($ch, CURLOPT_URL, $url);
        curl_setopt($ch, CURLOPT_HTTPHEADER, array(self::HTTPHEADER));
        curl_setopt($ch, CURLOPT_POST, true);
        curl_setopt($ch, CURLOPT_RETURNTRANSFER, true);
        curl_setopt($ch, CURLOPT_CONNECTTIMEOUT, 30);
        curl_setopt($ch, CURLOPT_USERAGENT, self::USERAGENT);
        curl_setopt($ch, CURLOPT_SSL_VERIFYPEER, false);
        curl_setopt($ch, CURLOPT_SSL_VERIFYHOST, false);
        curl_setopt($ch, CURLOPT_SSLCERT, self::CERTFILE);
        curl_setopt($ch, CURLOPT_SSLKEY, self::KEYFILE);
        curl_setopt($ch, CURLOPT_SSLKEYTYPE, "DER");
        curl_setopt($ch, CURLOPT_SSLCERTPASSWD, self::SERTPASS);

        if ($this->debug) {
            $this->logout = fopen(self::TMPLOGFILE, 'w+');
            curl_setopt($ch, CURLOPT_VERBOSE, TRUE);
            curl_setopt($ch, CURLOPT_STDERR, $this->logout);
        }

        if ($data) {
            $data = is_array($data) ? http_build_query($data) : $data;
            curl_setopt($ch, CURLOPT_POSTFIELDS, $data);
        }
        return $ch;
    }

    public function execCurl($curlHandler)
    {
        $result = curl_exec($curlHandler);

        if ($this->debug) {
            $this->curlTrace = file_get_contents(self::TMPLOGFILE);
            unlink(self::TMPLOGFILE);
        }

        return $result;
    }

    public function doRequest($method, $data) {
        $this->lastMethod = $method;
        $this->lastRequestData = $data;
        $methodUrl = $this->getMethodUrl($method);
        $ch = $this->initCurl($methodUrl, $data);
        $result = $this->execCurl($ch);

        if (!$result) {
            $this->curlErrorHandler($ch);
            return FALSE;
        }

        $xml = $this->curlSuccessHandler($result);

        if ($this->YMErrorHandler($xml)) {
            return FALSE;
        }

        return $xml;
    }

    public function log($string)
    {
        $logfileHandler = fopen($this->getLogfilePath(), 'a+');
        fwrite($logfileHandler, "\n--\n");
        fwrite($logfileHandler, date('Y-m-d H:i:s') . "\n");
        fwrite($logfileHandler, $string . "\n");
        fclose($logfileHandler);
    }

    /**
     * @param Error $error
     */
    public function logError($error = NULL)
    {
        if (!$error && !$this->error) {
            return;
        }
        if (!$error) {
            $error = $this->error;
        }
        $this->log($error->dump());
    }

    /**
     * @param $result
     * @return SimpleXMLElement
     */
    private function curlSuccessHandler($result)
    {
        $this->error = NULL;
        $this->lastResponse = $result;
        $this->lastXML = new \SimpleXMLElement($result);

        return $this->lastXML;
    }

    private function curlErrorHandler($curlHandler)
    {
        $this->error = new Error(
            'curl',
            curl_errno($curlHandler),
            curl_error($curlHandler),
            array('requestData' => $this->getLastRequestDataString())
        );
        $this->logError();
        $this->log("CURL TRACE\n" . $this->curlTrace);
        $this->lastResponse = "";
    }

    private function YMErrorHandler($xml)
    {
        if ($xml['error'] == '0') {
            return FALSE;
        }

        $dataError = array(
            'status' => $xml['status'],
            'requestData' => $this->getLastRequestDataString()
        );

        $this->error = new Error(
            'ym_api',
            $xml['error'],
            $xml['techMessage'],
            $dataError
        );
        $this->logError();
        return TRUE;
    }

    private function getLastRequestDataString() {
        $pairs = array();
        foreach($this->lastRequestData as $key => $value) {
            $pairs[] = $key . ': ' . $value;
        }
        return implode(', ', $pairs);
    }
}

class Error
{

    public $type = "";
    public $message = "";
    public $code = "";
    public $data = NULL;

    /**
     * @param $type
     * @param $code
     * @param $message
     * @param mixed $data
     */
    function __construct($type, $code, $message, $data = NULL)
    {
        $this->type = $type;
        $this->code = $code;
        $this->message = $message;

        if($data) {
            $this->data = $data;
        }
    }

    function dump()
    {
        $dump = $this->type;
        $dump .= "; ";
        $dump .= "Number: $this->code; ";
        $dump .= $this->message . "; ";

        if ($this->data) {
            $dump .= "Data: \n";
            $dump .= $this->dumpData();
        }

        return $dump;
    }

    function dumpData($data = null) {
        $data = $data ? $data : $this->data;

        if (is_array($this->data)) {
            $dumpArray = array();
            foreach($this->data as $key => $value) {
                $dumpArray[] = $key . '=' . $value;
            }
            return implode(', ', $dumpArray);
        }

        return str_replace("\n", " ", print_r($this->data, true));
    }
}
?>
